### PR TITLE
Implement null connection

### PR DESCRIPTION
**Pull request by GawainLynch**
- Add Genkgo\Mail\Protocol\NullConnection class
- Be more informative when throwing exception on unsupported scheme

Use:
```
ClientFactory::fromString('null://localhost/')->newClient()
```

Use cases:
- Testing
- Default set up parameters
- Maintenance mode

Please feel very free to throw rejection, requests for changes, or merge my way. :wink:

### CODE REVIEW

**Comment by frederikbosch** [comment]
@GawainLynch Thanks! Will look at this next week. Although I have one remark already, the `NullConnection` would belong in the `Smtp` namespace since it is dedicated to SMTP only.

**Comment by GawainLynch** [comment]
> `NullConnection` would belong in the `Smtp` namespace since it is dedicated to SMTP only.

I'm going to blame that on "Friday night" … I'll dig back into it in the morning. :+1:

**Comment by frederikbosch** [comment]
And tests for the class are missing.

**Comment by GawainLynch** [comment]
Moved NS & added tests.

Just to note, I had left out tests on the original PR as I had (originally) copied `PlainTcpConnection` and that was marked with `@codeCoverageIgnore` … plus this is itself a test class :wink:

**Comment by frederikbosch** [comment]
Thanks. I’d love to have tests though. There is defined behaviour, which is pretty well testable. This is not the case for the PlainTcpConnection, because you will have to open a tcp connection in another thread (other php process). Therefore I marked it as ignored for code coverage.

**Comment by GawainLynch** [comment]
Not a problem in the world for me. I'm just in the process of "getting to know how you work" … every project does things (slightly or more) different.

**Comment by frederikbosch** [comment]
Oh, I understand completely, and I am very happy with the fact you are contributing! In terms of how I work regarding to code: as strict as possible ;)

**Comment by GawainLynch** [comment]
Actually, on that … could you ping me via email (address on my GH profile), I have a couple of things to quickly run by you.

File: test/Unit/Protocol/Smtp/NullConnectionTest.php [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        self::assertSame($expect, $receive);
+    }
+
+    /**
+     * @test
+     */
+    public function it_does_nothing_when_upgrade_called(): void
+    {
+        $connection = new NullConnection();
+        $connection->upgrade(42);
```
Comment (frederikbosch) [inline test/Unit/Protocol/Smtp/NullConnectionTest.php]: Maybe use a constant here.

File: src/Protocol/Smtp/NullConnection.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    {
+    }
+
+    public function disconnect(): void
+    {
+    }
+
+    public function send(string $request): int
+    {
+        $this->setReceive($request);
```
Comment (frederikbosch) [inline src/Protocol/Smtp/NullConnection.php]: My aversion to `set` is pretty big... :) I'd prefer something like `pushResponseToBuffer`.

File: src/Protocol/Smtp/ClientFactory.php [hunk 3]
```diff
@@ -224,10 +224,14 @@ public static function fromString(string $dataSourceName):ClientFactory
                     $components['port'] ?? 25
                 );
                 break;
+            case 'null':
```
Comment (frederikbosch) [inline src/Protocol/Smtp/ClientFactory.php]: I don't know if we should add `null://` to the client factory. A new `NullConnection` is not something that typically requires a factory. `new Client(new NullConnection())` is not harder than `ClientFactory::fromString()->newClient()`. Moreover, `NullConnection` is typically required during testing. So there is no argument in adding it for the developer's easiness. Unless, you have sound reasons to have it in the factory, I'd say we do not add it to the factory.

Comment (GawainLynch) [reply src/Protocol/Smtp/ClientFactory.php]: > […] is not harder than […]

That is both fair and correct …

> Unless, you have sound reasons to have it in the factory, I'd say we do not add it to the factory.

My counter — that I am only adding for historical reference — is that my aversion to inconsistency is approximate to your aversion to setters :wink: 

However, personal ideology shouldn't stand in the way of a practical solution … so consider it "removed" :+1:

File: src/Protocol/Smtp/NullConnection.php (hunk 2 above)
Comment (GawainLynch) [reply src/Protocol/Smtp/NullConnection.php]: Done

File: test/Unit/Protocol/Smtp/NullConnectionTest.php (hunk 1 above)
Comment (GawainLynch) [reply test/Unit/Protocol/Smtp/NullConnectionTest.php]: Done

File: src/Protocol/Smtp/ClientFactory.php (hunk 3 above)
Comment (GawainLynch) [reply src/Protocol/Smtp/ClientFactory.php]: Oh, and linking #10 as when that gets moving, I'll add your suggestion to the docs.

**Comment by frederikbosch** [comment]
Much obliged!

**Comment by GawainLynch** [comment]
A pleasure :smiley_cat:

**Pull request merged**